### PR TITLE
SONARJAVA-1470 Inappropriate regular expressions

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      String regexp = LiteralUtils.trimQuotes(((LiteralTree) firstArg).value());
+      return regexp.matches(INAPPROPRIATE_REGEXPS);
+    }
+    return false;
+  }
+
+  private static boolean isFileSeparator(ExpressionTree firstArg) {
+    if (firstArg.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) firstArg;
+      return mse.identifier().name().equals("separator") && mse.expression().symbolType().is("java.io.File");
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: MAJOR Move the "separator" string literal on the left side of this string comparison. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1132)

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.java.tag.Tag;
+import org.sonar.plugins.java.api.tree.*;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  private static boolean isInappropriateRegexpStringLiteral(ExpressionTree firstArg) {
+    if (firstArg.is(Tree.Kind.STRING_LITERAL)) {
+      String regexp = LiteralUtils.trimQuotes(((LiteralTree) firstArg).value());
+      return regexp.matches(INAPPROPRIATE_REGEXPS);
+    }
+    return false;
+  }
+
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: MAJOR Move the "separator" string literal on the left side of this string comparison. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1132)

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.collect.ImmutableList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.java.tag.Tag;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: MAJOR Explicitly import the specific classes needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2208)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      String regexp = LiteralUtils.trimQuotes(((LiteralTree) firstArg).value());
+      return regexp.matches(INAPPROPRIATE_REGEXPS);
+    }
+    return false;
+  }
+
+  private static boolean isFileSeparator(ExpressionTree firstArg) {
+    if (firstArg.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) firstArg;
+      return mse.identifier().name().equals("separator") && mse.expression().symbolType().is("java.io.File");
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: MAJOR Move the "separator" string literal on the left side of this string comparison. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1132)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodMatcher.create().typeDefinition("java.lang.String").name("replaceAll").withNoParameterConstraint()
+      );
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree firstArg = mit.arguments().get(0);
+    if (isInappropriateRegexpStringLiteral(firstArg) || isFileSeparator(firstArg)) {
+      reportIssue(firstArg, "");
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: no message?!

Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java]: The message is missing.

**Pull request merged**